hkmc2/shared/src/test/mlscript/ucs/examples/EitherOrBoth.mls

```diff
@@ -91,3 +91,9 @@ fun eobToString[A, B](eob: EitherOrBoth[A, B]): Str =
     Left(left) then "Left(" ~ Str.from(left) ~ ")"
     Right(right) then "Right(" ~ Str.from(right) ~ ")"
     Both(left, right) then "Both(" ~ Str.from(left) ~ ", " ~ Str.from(right) ~ ")"
+//│ FAILURE: Unexpected type error
+//│ FAILURE LOCATION: resolve (ImplicitResolver.scala:122)
+//│ ╔══[ERROR] Unexpected moduleful reference.
+//│ ║  l.89: 	fun eobToString[A, B](eob: EitherOrBoth[A, B]): Str =
+//│ ║        	                                                ^^^
+//│ ╙── Function must be marked as 'module' in order to have a module signature.
```

Comment:
Here's a problem with overloaded definitions. There is a `Str` class definition in the prelude and a `Str` module in `mlscript-compile/Str.mls`. What I expect is that there is a `BlockMemberSymbol` of the two `Str`, but currently there are two `BlockMemberSymbol` and the imported one from `mlscript-compile/Str.mls` shadows the one from prelude. The consequence here (and in some other test cases) is that `Str` is recognized as a module instead of a class. If it were one `BlockMemberSymbol` containing one `ClassSymbol` and one `ModuleSymbol`, then we should be able to tell which kind of symbol has priority here.

Reply:
That's not a problem with overloaded definitions, though. Just conflicting imports/builtins. Overloads must be defined in the very same scope, otherwise it's madness.

Reply:
I see. Then I'll have to rename the `Str` module in `mlscript-compile/Str.mls`. I guess it should be an error to define a module named `Str` at the very first place where it is defined.

Reply:
But we still have to deal with overloaded definitions. A symbol used to be recognized as a module as long as one of its overloaded definition is a symbol. 

I think it should be more clever in the sense that it should be able to know if the user wants a class type or module type:

```fs
class Str
module Str

// a module
fun f: module Str

// a class
fun f: Str
```

How do you think about this behavior? We will still perform checks to prevent user from really returning a modueful or non-moduleful term that doesn't match the signature.

Reply:
Yes, that's the intended behavior. `module Str` disambiguates the overloaded definition by selecting the module part. Same as other constructs, when not referred to in their expected context. For example, in
```
class Foo
fun Foo
```
then `f(Foo)` refers to the `fun`, but `f(class Foo)` refers to the class (using it as a first-class value).

Reply:
I've been wanting to add `Disambiguate(term, kind)(innerSymbol)` nodes to `Term` which would make it perfectly clear what we're looking at. Just haven't had time yet.

Reply:
> I see. Then I'll have to rename the `Str` module in `mlscript-compile/Str.mls`.

Note that eevntually `Str` should just refer to a module with helper functions along with an overloaded type declaration, probably both defined in Predef. `Str` actually has nothing to do in `Prelude`. We can fix these details later, when overloaded is supported.

> I guess it should be an error to _define_ a module named `Str` here.

I don't see why. We don't want to (and reasonably can't) prevent people from defining modules with the names they choose.

Reply:
> I don't see why. We don't want to (and reasonably can't) prevent people from defining modules with the names they choose.

You said it is "conflicting". So I guess this should be an error? (or shadowing the previous definition should be the intended behavior?)

Given that you said eventually `Str` should be a module with helper functions, I guess currently let's just define it to be a `object` so that it doesn't fail the module checks.

Reply:
Yes that's a fine plan for now.

> You said it is "conflicting". So I guess this should be an error? (or shadowing the previous definition should be the intended behavior?)

Yes, it's a name conflict. No, it shouldn't be an error to name something `Str`, obviously. And yes, if you import something that shadows something else, you shouldn't make a surprised pikatchu face when the old symbol is shadowed.